internal/cluster/certificates.go

```diff
@@ -0,0 +1,411 @@
+	bootstrapv1 "sigs.k8s.io/cluster-api-bootstrap-provider-kubeadm/api/v1alpha2"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha2"
+	"sigs.k8s.io/cluster-api/util/certs"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	rootOwnerValue = "root:root"
+
+	// TLSKeyDataName is the name of the secret key within the certificate secret of the key
+	TLSKeyDataName = "tls.key"
+
+	// TLSCrtDataName is the name of the secret key within the certificate secret of the certificate
+	TLSCrtDataName = "tls.crt"
+
+	// ClusterCAName is the secret name suffix for apiserver CA
+	ClusterCAName = "ca"
+
+	// EtcdCAName is the secret name suffix for the Etcd CA
+	EtcdCAName = "etcd"
+
+	// ServiceAccountName is the secret name suffix for the Service Account keys
+	ServiceAccountName = "sa"
+
+	// FrontProxyCAName is the secret name suffix for Front Proxy CA
+	FrontProxyCAName = "proxy"
+)
+
+var (
+	// ErrMissingCertificate is an error indicating a certificate is entirely missing
+	ErrMissingCertificate = errors.New("missing certificate")
+
+	// ErrMissingCrt is an error indicating the crt file is missing from the certificate
+	ErrMissingCrt = errors.New("missing crt data")
+
+	// ErrMissingKey is an error indicating the key file is missing from the certificate
+	ErrMissingKey = errors.New("missing key data")
+)
+
+// Certificates are the certificates necessary to bootstrap a cluster.
+type Certificates []*Certificate
+
+// NewCertificates return an initialized but empty set of CA certificates needed to bootstrap a cluster.
+func NewCertificates() Certificates {
+	return Certificates{
+		&Certificate{Name: ClusterCAName},
+		&Certificate{Name: EtcdCAName},
+		&Certificate{Name: ServiceAccountName},
+		&Certificate{Name: FrontProxyCAName},
+	}
+}
+
+// GetByName returns a certificate by the given name.
+// This could be removed if we use a map instead of a slice to hold certificates, however other code becomes more complex.
+func (c Certificates) GetByName(name string) *Certificate {
+	for _, certificate := range c {
+		if certificate.Name == name {
+			return certificate
+		}
+	}
+	return nil
+}
+
+// Lookup looks up each certificate from secrets and populates the certificate with the secret data.
+func (c Certificates) Lookup(ctx context.Context, ctrlclient client.Client, cluster *clusterv1.Cluster) error {
+	// Look up each certificate as a secret and populate the certificate/key
+	for _, certificate := range c {
+		secret := &corev1.Secret{}
+		key := client.ObjectKey{
+			Name:      certificate.SecretName(cluster.Name),
+			Namespace: cluster.Namespace,
+		}
+		if err := ctrlclient.Get(ctx, key, secret); err != nil {
+			if apierrors.IsNotFound(err) {
+				continue
+			}
+			return errors.WithStack(err)
+		}
+		// If a user has a badly formatted secret it will prevent the cluster from working.
+		kp, err := secretToKeyPair(secret)
+		if err != nil {
+			return err
+		}
+		if kp == nil {
+			return errors.New("SECRET IS NIL")
```

Comment:
😄

Reply:
... SHhhhhhh you didn't see annyyyything